rfc/rfc-73/rfc-73.md

```diff
@@ -0,0 +1,416 @@
+
+However, the current implementation cannot be extended as-is to support atomic operations across multiple tables. First
+of all, we need a notion of a "database" and its tables to be able to associate a transaction with multiple tables.
+Secondly, Hudi's timeline would need to evolve to account for changes across multiple tables. This introduces
+complexities in tracking and managing the order of operations across tables. With multiple tables involved, the points
+of potential failures increase. A failure in one table can cascade and affect the entire transaction. In case of
+failures, rolling back changes becomes more intricate in multi-table scenarios. Ensuring data consistency across tables
+during rollbacks, especially when there are inter-table dependencies, introduces additional challenges. Finally, the
+current concurrency control implementation is not designed to handle multi-table transactions. Hence, this RFC proposes
+a new design to support multi-table transactions.
+
+## Design
+
+First of all, let us discuss the goals and non-goals which will help in understanding the design considerations better.
+
+### Goals
+
+1. **Atomicity Across Tables:** Ensure that a set of changes spanning multiple tables either fully completes or fully
+   rolls back. No partial commits should occur.
+2. **Consistency:** Ensure that, after every transaction, all involved tables are in a consistent state. Avoid dirty
+   reads, non-repeatable reads, handle read-write and write-write conflicts.
+3. **Isolation:** Multiple concurrent transactions should not interfere with each other. One transaction shouldn't see
+   the intermediate states of another ongoing transaction.
+4. **Durability:** Once a multi-table transaction is committed, the changes should be permanent and recoverable, even
+   after system failures.
+5. **Performance:** The multi-table transaction mechanism should introduce minimal overhead. Its performance impact on
+   typical Hudi operations should be kept low. By corollary, locking (in case of OCC) cannot be coarser than file level.
+6. **Monitoring:** Integration with tools such as hudi-cli to track the status and health of transactions. Separate
+   transaction metrics. Also, provide comprehensive logs for debugging transaction failures or issues.
+7. **Integration with Current Hudi Features:** The multi-table transaction should be seamlessly integrated with existing
+   Hudi features and should not break existing APIs or workflows. In
+   particular, `TransacationManager`, `LockManager`, `ConflictResolutionStrategy` APIs should work as they do today with
+   single table.
+8. **Configurability:** Allow users to configure the behavior of multi-table transactions, e.g., concurrency control
+   mechanism, timeout durations, etc.
+9. **Recovery**: Rollbacks, savepoint and restore should be supported.
+
+### Non-goals
+
+1. **Cross-Database Transactions:** Transactions spanning multiple databases might be too complex as an initial target.
```

Comment:
see comments above. I understand what you are getting at here, but we could debate if its sufficient to have txns spanning a single database (or may be give it a different name e.g 'transaction context', since its not really tied to databases to begin with). but limiting to "database" seems like the most intuitive. Kind of still mulling this point.

Reply:
We need to have some notion of boundary. We can call it database, or namespace, or lake. As you said, database seems most intuitive. It's also easily mapped to catalog. One can list all databases in a catalog. 

A database will be contained within a lake, and a lake can have more than 1 database. The current proposal is to just limit the scope of transaction to database level. For example, in medallion architecture, you have bronze, silver and gold lakes and each one could have databases depending on usecase (e.g. clickstream, postgres cdc, etc). A transaction can only span across tables within a database.
Do you see some usecases where it would be useful to have cross-database or cross-lake transactions?

Reply:
Agree we need to define some boundary. we can debate if we give this a different name than database.